src/main/java/com/microsoft/azure/relay/ClientWebSocket.java

```diff
@@ -200,60 +201,60 @@ boolean isOpen() {
 	 * @throws TimeoutException Throws when the sending task does not complete within the given timeout.
 	 */
 	public CompletableFuture<Void> writeAsync(Object data, Duration timeout) {
-		return writeAsync(data, timeout, WriteMode.BINARY);
+		return writeAsync(data, timeout, true, WriteMode.BINARY);
 	}
 	
 	/**
 	 * Sends the data to the remote endpoint within a timeout in one of the WriteModes.
 	 * 
 	 * @param data Message to be sent.
 	 * @param timeout The timeout to connect to send the data within. May be null to indicate no timeout limit.
+	 * @param isEnd Indicates if the data sent is the end of a message
 	 * @param mode The type of the message to be sent.
 	 * @return A CompletableFuture which completes when websocket finishes sending the bytes.
 	 * @throws TimeoutException Throws when the sending task does not complete within the given timeout.
 	 */
-	CompletableFuture<Void> writeAsync(Object data, Duration timeout, WriteMode mode) {
+	CompletableFuture<Void> writeAsync(Object data, Duration timeout, boolean isEnd, WriteMode mode) {
 		if (this.isOpen()) {
 			if (data == null) {
 				// TODO: Log warns sending nothing because message is null
 				return CompletableFuture.completedFuture(null);
 			}
 			else {
-				return CompletableFutureUtil.timedRunAsync(timeout, () -> {
-					RemoteEndpoint.Basic remote = this.session.getBasicRemote();
-					
-					try {
-						if (mode.equals(WriteMode.TEXT)) {
-							String text = data.toString();
-							RelayLogger.logEvent("writingBytes", this, "text");
-							Writer writer = remote.getSendWriter();
-							writer.write(text);
-							writer.close();
-							RelayLogger.logEvent("doneWritingBytes", this, String.valueOf(text.length()));
-						}
-						else {
-							ByteBuffer bytes = null;
-							
-							RelayLogger.logEvent("writingBytes", this, "binary");
-							if (data instanceof byte[]) {
-								bytes = ByteBuffer.wrap((byte[]) data);
-							} 
-							else if (data instanceof ByteBuffer) {
-								bytes = (ByteBuffer) data;
+				RemoteEndpoint.Basic remote = this.session.getBasicRemote();
+				RelayLogger.logEvent("writingBytes", this, mode.toString());
+				
+				// The websocket API will throw if multiple sends are attempted on the same websocket simultaneously
```

Comment:
>The websocket API will throw if multiple sends are attempted on the same websocket simultaneously [](start = 7, length = 97)

I'm curious if it's the libraries job to ensure users don't do this bad thing or the user's responsibility. My inclination is that if we try to do it for them then they don't get any guarantees as to the order of their data. If we don't do it for them, then they must wait until the previous write finishes before making the next write and their data is actually guaranteed to be in the right order.

Basically this doesn't guarantee ordering:
webSocket.writeAsync( ...) ; // bytes 0-9, NO WAIT
webSocket.writeAsync(...); // bytes 10-19, NO WAIT

Reply:
Or is it to solve the issue where user calls writeAsync the closeAsync without waiting?

CF writeFuture = ws.writeAsync(...);
CF closeFuture = ws.closeAsync(); // close waits for inflight send to drain?

---
In reply to: [291712180](https://github.com/Azure/azure-relay-java/pull/45#discussion_r291712180) [](ancestors = 291712180)

Reply:
This lock is to guard the websocket from multiple writes concurrently, not related to close.

All the unit tests are doing synchronized writes as of now and not running into this issue. I will do more investigation on this later in a separate PR. Probably going to leave this lock out for now.

---
In reply to: [292064317](https://github.com/Azure/azure-relay-java/pull/45#discussion_r292064317) [](ancestors = 292064317,291712180)

Reply:
If a user calls writeAsync twice concurrently I think that ideally we should give them an exception. Even if no exception it's pretty much into undefined territory.

**However**, if our timer based flush is racing with the user calling another writeAsync we almost certainly need to serialize things there.

---
In reply to: [292622840](https://github.com/Azure/azure-relay-java/pull/45#discussion_r292622840) [](ancestors = 292622840,292064317,291712180)